### PR TITLE
Interleaved Multi-head Attention Operators

### PR DESCRIPTION
**Pull request by blchu**
Replaced batched dot product in multi-head attention with interleaved_matmul attention operators to improve performance. Also changes the batch-major data to time-major format while in the model to comply with the new operator requirements.

Replacing https://github.com/awslabs/sockeye/pull/814 due to issues with diffs against the master branch.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by blchu** [comment]
@tdomhan My changes are not backward compatible with previous versions of Sockeye, and so requires a major version update. How should I modify the backwards compatibility test that you introduced to account for this?

**Comment by tdomhan** [comment]
That's a good point! I guess there's no way to make things backwards compatible. 

So what we should do is create a `model_2.2.x` using the sequence copy data/the command from here:
https://github.com/awslabs/sockeye/blob/master/test/data/model_2.1.x/README.md
and delete model_2.1.x.

Does that work for you?

**Comment by blchu** [comment]
Sure, I'll go ahead and do that.

File: sockeye/layers.py [hunk 1]
```diff
@@ -385,8 +352,7 @@ def __init__(self,
     def _attend(self,
                 F,
                 queries: mx.sym.Symbol,
-                keys: mx.sym.Symbol,
-                values: mx.sym.Symbol,
+                key_values: mx.sym.Symbol,
```
Comment (barbaradarques) [inline sockeye/layers.py:355]: Update docstring accordingly.

File: sockeye/layers.py [hunk 2]
```diff
@@ -587,29 +521,14 @@ def __init__(self,
...
+            self.ff_kv = quantization.QuantizableDense(in_units=depth_key_value, units=2*depth_att, flatten=False, use_bias=False, prefix='kv2h_', dtype=dtype)
 
     def hybrid_forward(self, F,
                        queries: mx.sym.Symbol,
                        memory: mx.sym.Symbol,
                        memory_lengths: Optional[mx.sym.Symbol] = None,
                        bias: Optional[mx.sym.Symbol] = None,
-                       projected_memory_keys: Optional[mx.sym.Symbol] = None,
-                       projected_memory_values: Optional[mx.sym.Symbol] = None) -> mx.sym.Symbol:  # mypy: ignore
+                       projected_memory_kv: Optional[mx.sym.Symbol] = None) -> mx.sym.Symbol:  # mypy: ignore
```
Comment (barbaradarques) [inline sockeye/layers.py:531]: Update docstring.

File: sockeye/layers.py [hunk 3]
```diff
@@ -325,9 +293,9 @@ def cast(self, dtype):
         self._dtype = dtype
         super().cast(dtype)
 
-    def hybrid_forward(self, F, queries, keys, values, lengths=None, bias=None):
-        # (n, lq, lk)
-        logits = F.batch_dot(lhs=queries, rhs=keys, transpose_b=True)
+    def hybrid_forward(self, F, queries, key_values, heads, lengths=None, bias=None):
```
Comment (barbaradarques) [inline sockeye/layers.py:296]: Update calls/docstrings at `ProjectedDotAttention` (https://github.com/awslabs/sockeye/pull/884/files#diff-74c804465509cbab074609c50c13a8f7L710) and `PlainDotAttention` (https://github.com/awslabs/sockeye/pull/884/files#diff-74c804465509cbab074609c50c13a8f7L642).

File: sockeye/transformer.py [hunk 4]
```diff
@@ -226,13 +217,12 @@ def hybrid_forward(self, F,
                        source: mx.sym.Symbol,
                        source_bias: mx.sym.Symbol,
                        autoregr_states: mx.sym.Symbol,
-                       enc_att_k: Optional[mx.sym.Symbol] = None,
-                       enc_att_v: Optional[mx.sym.Symbol] = None) -> Tuple[mx.sym.Symbol,
+                       enc_att_kv: Optional[mx.sym.Symbol] = None) -> Tuple[mx.sym.Symbol,
                                                                            mx.sym.Symbol]:
-        target_autoregr, *new_autoregr_states = self.autoregr_layer(self.pre_autoregr_layer(target, None),
```
Comment (barbaradarques) [inline sockeye/transformer.py:232]: As mentioned above, we should maintain the unpacking in order to keep supporting autoregressive layers with more than one state tensor.

**Review by barbaradarques — CHANGES_REQUESTED** [review]

**Comment by blchu** [comment]
@tdomhan @barbaradarques @mjdenkowski @fhieber Are there any potential issues left with this PR, or would it be ready to be merged in now?

**Comment by barbaradarques** [comment]
Hi @blchu, I'm still finding a discrepancy in perplexity, both with and without SSRUs.

**Comment by barbaradarques** [comment]
But thanks for iterating! All the things I pointed out in the previous review were covered.

**Comment by barbaradarques** [comment]
Updating on the output of my tests, I also got an output of only <unk>'s in the SSRU version.

**Comment by blchu** [comment]
@barbaradarques could you tell me what you ran to get those results so I can try to replicate it on my side? From looking at system test outputs I'm not noticing any significant discrepancy in perplexity, and training both a transformer model and SSRU model on a WMT dataset (and comparing to master) gets expected results for training perplexity, BLEU, and translation output.

**Comment by barbaradarques** [comment]
@blchu Interesting. I will double check my own tests then, before we deep diver into this.

**Comment by barbaradarques** [comment]
Also, in order to remove the effect of the Xavier initializer, I replaced it with [mxnet.initializer.One](https://mxnet.apache.org/versions/1.6/api/python/docs/api/initializer/index.html#mxnet.initializer.One).

**Comment by fhieber** [comment]
I tried reproducing the differences locally on my laptop, using a common toy sequence copy task (validation data == training data). The command is:
```
python3 -m sockeye.train --use-cpu -s data/dev.source -t data/dev.target -vs data/dev.source -vt data/dev.target --num-embed 32 --batch-size 10 --batch-type sentence -o blchu_transformer --num-layers 2:2 --overwrite-output --transformer-model-size 32 --weight-tying-type src_trg_softmax --max-updates 8000 --decode-and-evaluate 10 --checkpoint-interval 2000 \
	--decoder DECODER -o MODEL
```
where DECODER is `{transformer, ssru_transformer}` and MODEL is an output folder.

The training perplexity after the first 50 batches:
```
blchu_ssru/log: Epoch[0] Batch [50]	perplexity=19.079991
blchu_transformer/log: Epoch[0] Batch [50]	perplexity=18.648705
master_ssru/log: Epoch[0] Batch [50] perplexity=18.851622
master_transformer/log: Epoch[0] Batch [50]	perplexity=17.698212
```

After 8000 updates:
```
blchu_ssru/log: Epoch[79] Batch [8000] perplexity=2.035750
blchu_transformer/log: Epoch[79] Batch [8000]perplexity=2.099237
master_ssru/log: Epoch[79] Batch [8000]	perplexity=2.030470
master_transformer/log: Epoch[79] Batch [8000]perplexity=2.107919
```

Last checkpoint (4):
```
blchu_ssru/log: Checkpoint [4]	Validation-perplexity=1.754443	Validation-bleu=1.000000	Validation-chrf=1.000000	Validation-rouge1=0.930180	Validation-rouge2=0.937282	Validation-rougel=0.943623	Validation-length-ratio-mse=1.000000	Validation-avg-sec-per-sent=0.016261	Validation-decode-walltime=0.162612
blchu_transformer/log: Checkpoint [4]	Validation-perplexity=1.758212	Validation-bleu=1.000000	Validation-chrf=1.000000	Validation-rouge1=0.930180	Validation-rouge2=0.937282	Validation-rougel=0.943623	Validation-length-ratio-mse=1.000000	Validation-avg-sec-per-sent=0.022654	Validation-decode-walltime=0.226536
master_ssru/log: Checkpoint [4]	Validation-perplexity=1.754248	Validation-bleu=1.000000	Validation-chrf=1.000000	Validation-rouge1=0.930180	Validation-rouge2=0.937282	Validation-rougel=0.943623	Validation-length-ratio-mse=1.000000	Validation-avg-sec-per-sent=0.016961	Validation-decode-walltime=0.169605
master_transformer/log: Checkpoint [4]	Validation-perplexity=1.756007	Validation-bleu=1.000000	Validation-chrf=1.000000	Validation-rouge1=0.930180	Validation-rouge2=0.937282	Validation-rougel=0.943623	Validation-length-ratio-mse=1.000000	Validation-avg-sec-per-sent=0.032418	Validation-decode-walltime=0.324185
```


This looks fine to me, @barbaradarques

**Comment by barbaradarques** [comment]
I probably tampered with it too much. So, other than that, everything else looks good to me!

**Review by barbaradarques — APPROVED** [review]

**Review by fhieber — APPROVED** [review]
We've gone over this a lot, thanks for iterating and bringing it over the finish line. LGTM

**Review by tdomhan — APPROVED** [review]

**Pull request merged**